### PR TITLE
using node

### PR DESCRIPTION
**Pull request by vlandham**
added a section on using node. See what you think!

### CODE REVIEW

File: public/src/node.js [hunk 1]
```diff
@@ -0,0 +1,146 @@
+// # Analyzing Data with Node
+//
+// As mentioned in the [introduction](getting_started.html), this guide is mostly geared for client-side data analysis, but with a few augmentations, the same tools can be readily used server-side with [Node](https://nodejs.org/).
+//
+// And the data is too large, this might in fact be your only option if you want to use JavaScript for your data analysis. None of your users will wait 5 minutes on your site while a giant file is loaded up, no matter how cool the analysis.
```
Comment (iros) [inline public/src/node.js]: "And the data is..." -> "If the data is..."

File: public/node.html [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+        
+        <li id="section-1">
+            <div class="annotation">
+              
+              <div class="pilwrap ">
+                <a class="pilcrow" href="#section-1">&#182;</a>
+              </div>
+              <h1 id="analyzing-data-with-node">Analyzing Data with Node</h1>
+<p>As mentioned in the <a href="getting_started.html">introduction</a>, this guide is mostly geared for client-side data analysis, but with a few augmentations, the same tools can be readily used server-side with <a href="https://nodejs.org/">Node</a>.</p>
+<p>And the data is too large, this might in fact be your only option if you want to use JavaScript for your data analysis. None of your users will wait 5 minutes on your site while a giant file is loaded up, no matter how cool the analysis.</p>
```
Comment (tafsiri) [inline public/node.html]: missing 'if' before 'the data is too large'

File: public/src/node.js (hunk 1 above)
Comment (iros) [inline public/src/node.js]: "None of your users will wait 5 minutes on your site while a giant file is loaded up, no matter how cool the analysis."

I know what you're saying here: If your data is too large, processing it client side might incur too long of a wait. However, you can use node.js to build servers, and then try to do the same data analysis INSIDE YOUR CALLBACKS. Which, of course, is a bad idea. I'm not sure exactly how to rephrase it, but I think pointing that out might be worthwhile. Suggestion:

"Trying to deal with large data in the browser might result in your users having to wait for a long time. No user will wait for 5 minutes with a frozen browser, no matter how cool the analysis might be." <- More browser specific?

File: public/src/node.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+// To get started with Node, ensure both node and [npm](https://www.npmjs.com/), the Node package manager, are installed and available via the command line:
+//
+// ```bash
+// which node
+// # /usr/local/bin/node
+// which npm
+// # /usr/local/bin/npm
+// ```
+//
+// Your paths may be different then mine, but as long as `which` returns something, you should be good to go.
```
Comment (iros) [inline public/src/node.js:18]: Maybe add a sentence with a link to node.js's install instructions if they don't?

File: public/src/node.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+// ```bash
+// which node
+// # /usr/local/bin/node
+// which npm
+// # /usr/local/bin/npm
+// ```
+//
+// Your paths may be different then mine, but as long as `which` returns something, you should be good to go. 
+//
+// Create a new directory for your data analysis project. In this example, we have a directory with a sub-directory called `data` which contains our `animals.tsv` file inside.
```
Comment (iros) [inline public/src/node.js:22]: I think it's worth showing the full directory structure in a code block, like this:

```
animals
  data
    animals.tsv
```

File: public/node.html [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+<p>Next, we want to install our JavaScript tools, D3 and lodash. With Node, we can automate the process by using <code>npm</code>. Inside your data analysis directory run the following:</p>
+<pre><code class="lang-bash">npm install d3
+npm install lodash
+</code></pre>
+<p>You can see that npm creates a new sub-directory called <code>node_modules</code> by default, where your packages are installed. Everything is kept local, so you don’t have to worry about problems with missing or out-of-date packages. Your analysis tools for each project are ready to go.</p>
+<h2 id="requiring-modules">Requiring Modules</h2>
+<p>Now we create a separate JavaScript file to do our analysis in:</p>
+<pre><code class="lang-bash">touch analyze.js
+</code></pre>
+<p>Inside this file, we first <a href="http://openmymind.net/2012/2/3/Node-Require-and-Exports/">require</a> our external dependencies. With Node, everything is explicit. You need to indicate exactly which tools you want to use.</p>
```
Comment (tafsiri) [inline public/node.html]: I wouldn't say that 'require' makes node more explicit. You still have to load files in client side js explicitly. You could say it is synchronous and guaranteed to finish before moving on to the next line or something. But i wasn't sure what you were trying to say.

File: public/src/node.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+// ## Installing Node Modules
+//
+// Next, we want to install our JavaScript tools, D3 and lodash. With Node, we can automate the process by using `npm`. Inside your data analysis directory run the following:
+//
+// ```bash
+// npm install d3
+// npm install lodash
+// ```
+//
+// You can see that npm creates a new sub-directory called `node_modules` by default, where your packages are installed. Everything is kept local, so you don't have to worry about problems with missing or out-of-date packages. Your analysis tools for each project are ready to go.
```
Comment (iros) [inline public/src/node.js:41]: Is it worth mentioning package.json?

File: public/src/node.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+//
+// We also require the [file system](https://nodejs.org/api/fs.html) module. As we will see in a second, we need this to load our data - which is really the key difference between client-side and server-side use of these tools
+//
+// ## Loading Data in Node
+//
+// D3's data loading functionality is based on [XMLHttpRequest](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest), which is great, but Node does not have `XMLHttpRequest`. There are packages around this mismatch, but a more elegant solution is to just use Node's built in file system functionality to load the data, and then D3 to parse it.
+//
+// ```js
+// fs.readFile("data/animals.tsv", "utf8", function(error, data) {
+// data = d3.tsv.parse(data);
```
Comment (iros) [inline public/src/node.js]: Can we get indentation on these? I think we've done that in other sections.

File: public/node.html [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+bigAnimalsString = d3.csv.format(bigAnimals);
+
+fs.writeFile(<span class="hljs-string">"big_animals.csv"</span>, bigAnimalsString, <span class="hljs-function"><span class="hljs-keyword">function</span><span class="hljs-params">(err)</span> </span>{
+  <span class="hljs-built_in">console</span>.log(<span class="hljs-string">"file written"</span>);
+});
+</code></pre>
+<p>Run this with the same <code>node analysis.js</code> and now you should have a lovely little <code>big_animals.csv</code> file in your directory. It even takes care of the headers for you:</p>
+<pre><code>name,type,avg_weight
+hippo,mammal,<span class="hljs-number">3400</span>
+</code></pre><p>Now even BIG data is no match for us - using the power of JavaScript!</p>
```
Comment (tafsiri) [inline public/node.html]: is BIG data a reference to big animals? :).

File: src/node.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+// ```
+//
+// Run this with the same `node analysis.js` and now you should have a lovely little `big_animals.csv` file in your directory. It even takes care of the headers for you:
+//
+// ```
+// name,type,avg_weight
+// hippo,mammal,3400
+// ```
+// 
+// Now even BIG data is no match for us - using the power of JavaScript!
```
Comment (iros) [inline src/node.js]: LOL. BIG DATA.

File: public/src/node.js (hunk 1 above)
Comment (vlandham) [inline public/src/node.js]: yes - great suggestion. Incorrectly, I have in my head "offline" through all this - but you are right. Server side isn't offline.

will use.

File: public/node.html (hunk 5 above)
Comment (vlandham) [inline public/node.html]: ok. i'll take it out. I guess i was comparing to something like ruby - where you get a lot of your modules included implicitly. and the ones you do require - you have no control over how they are dropped in. Node, like python, makes you spell out exactly what you want.

File: public/src/node.js (hunk 6 above)
Comment (vlandham) [inline public/src/node.js:41]: sure. maybe with a link on if they want to know more... i'll see if i can find something good.

File: public/src/node.js (hunk 7 above)
Comment (vlandham) [inline public/src/node.js]: yeah. not sure what happened. will add.

**Pull request merged**